addons/stock/models/stock_location.py

```diff
@@ -83,7 +83,26 @@ def write(self, values):
         if 'usage' in values and values['usage'] == 'view':
             if self.mapped('quant_ids'):
                 raise UserError(_("This location's usage cannot be changed to view as it contains products."))
-        return super(Location, self).write(values)
+        if 'usage' in values or 'scrap_location' in values:
+            sbr1 = self.should_bypass_reservation()
+
+        res = super(Location, self).write(values)
+
+        if 'usage' in values or 'scrap_location' in values:
+            # 'if usage' again because this part must be executed after the write()
+            sbr2 = self.should_bypass_reservation()
+            if sbr1 != sbr2:
+                reserved_quantities = self.env['stock.move.line'].search_count([
```

Comment:
Same comment as above. If you want to change the location type while there is any stock move existing from/to it, then you should probably create a new location in all cases. I think the check could be simpler.